Add a ControlNet model & pipeline (#2407)
huggingface/diffusers: docs/source/en/api/pipelines/stable_diffusion/controlnet.mdx

the License. You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
-->

# Text-to-Image Generation with ControlNet Conditioning

## StableDiffusionControlNetPipeline

ControlNet by [@lllyasviel](https://huggingface.co/lllyasviel) is a neural network structure to control diffusion models by adding extra conditions.

There are 8 pre-trained ControlNet models that were trained to condition the original Stable Diffusion model on different inputs, 
such as edge detection, scribbles, depth maps, semantic segmentations and more.

Using the pretrained models we can provide control images (for example, a depth map) to control Stable Diffusion text-to-image generation so that it follows the structure of the depth image and fills in the details.

The original codebase/paper can be found here: 
- [Code](https://github.com/lllyasviel/ControlNet)

> - [Paper](https://arxiv.org/abs/2302.05543)

Reviewer: Please refer to [this documentation](https://huggingface.co/docs/diffusers/api/pipelines/stable_diffusion/pix2pix) as a reference to see how we usually structure the pipeline documentation. 

In particular, "Overview" should be the first sub-section. Under that, we include the paper name, the abstract, and other resources.

Author: Thanks! Already applied in this update. https://github.com/huggingface/diffusers/pull/2407/commits/06bb1db313d1e3b87d82d662ae02245a6894f30f

## Available checkpoints

> ControlNet requires a *control image* in addition to the text-to-image *prompt*. 

Reviewer: ```suggestion
ControlNet requires a *control image* in addition to the text-to-image *prompt*. 
Each pretrained model is trained using a different conditioning method that requires different images for conditioning the generated outputs. For example, Canny edge conditioning requires the control image to be the output of a Canny filter, while depth conditioning requires the control image to be a depth map. See the overview and image examples below to know more.
```

Each pretrained model is trained using a different conditioning method that requires different conditioning images. For example, Canny edge conditioning requires the control image to be the output of a Canny filter, while depth conditioning requires the control image to be a depth map.
See the overview and image examples.

All checkpoints are converted from [lllyasviel/ControlNet](https://huggingface.co/lllyasviel/ControlNet).

### ControlNet + Stable Diffusion 1.5

| Model Name | Control Image Overview| Control Image Example | Generated Image Example |
|---|---|---|---|

> |[takuma104/control_sd15_canny](https://huggingface.co/takuma104/control_sd15_canny)<br/> *Trained with canny edge detection* | A monochrome image with white edges on a black background.|<a href="https://huggingface.co/takuma104/controlnet_dev/blob/main/gen_compare/control_images/converted/control_bird_canny.png"><img width="64" style="margin:0;padding:0;" src="https://huggingface.co/takuma104/controlnet_dev/resolve/main/gen_compare/control_images/converted/control_bird_canny.png"/></a>|<a href="https://huggingface.co/takuma104/controlnet_dev/resolve/main/gen_compare/output_images/diffusers/output_bird_canny_1.png"><img width="64" src="https://huggingface.co/takuma104/controlnet_dev/resolve/main/gen_compare/output_images/diffusers/output_bird_canny_1.png"/></a>|

Reviewer: We should transfer these checkpoints to an appropriate location. Cc: @patrickvonplaten

|[takuma104/control_sd15_seg](https://huggingface.co/takuma104/control_sd15_seg)<br/>*Trained with semantic segmentation*  |An [ADE20K](https://groups.csail.mit.edu/vision/datasets/ADE20K/)'s segmentation protocol image.|<a href="https://huggingface.co/takuma104/controlnet_dev/blob/main/gen_compare/control_images/converted/control_room_seg.png"><img width="64" src="https://huggingface.co/takuma104/controlnet_dev/resolve/main/gen_compare/control_images/converted/control_room_seg.png"/></a>|<a href="https://huggingface.co/takuma104/controlnet_dev/resolve/main/gen_compare/output_images/diffusers/output_room_seg_1.png"><img width="64" src="https://huggingface.co/takuma104/controlnet_dev/resolve/main/gen_compare/output_images/diffusers/output_room_seg_1.png"/></a> |

## Resources

- [Colab Notebook Example](https://colab.research.google.com/drive/1AiR7Q-sBqO88NCyswpfiuwXZc7DfMyKA?usp=sharing)

> - [controlnet_hinter](https://github.com/takuma104/controlnet_hinter): Image Preprocess Library for ControlNet

Reviewer: We usually don't add a separate section for resources like this.

Reviewer: Instead we add a section on "Available PipelinesP" (see [here](https://huggingface.co/docs/diffusers/api/pipelines/stable_diffusion/pix2pix#available-pipelines)). @AK391 do we have a Space for ControlNet that we could link here?

Author: I have added a section for "Available Pipelines". I have left my Colab example, but it is okay to delete it as soon as Space for ControlNet becomes available.

https://github.com/huggingface/diffusers/pull/2407/commits/06bb1db313d1e3b87d82d662ae02245a6894f30f

## Usage example

- Basic Example (Canny Edge)

The conditioning image is an outline of the image edges, as detected by a Canny filter. This is the example we'll use to control the generation:

> ![White on black edges detected on Vermeer's Girl with a Pearl Earring portrait](https://huggingface.co/takuma104/controlnet_dev/resolve/main/vermeer_canny_edged.png)

Reviewer: Open a PR to [this repository](https://huggingface.co/datasets/huggingface/documentation-images), particularly under the `diffusers` folder.

Reviewer: ```suggestion
[White on black edges detected on Vermeer's Girl with a Pearl Earring portrait](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/diffusers/vermeer_canny_edged.png)
```

```python
from diffusers import StableDiffusionControlNetPipeline
from diffusers.utils import load_image

# Canny edged image for control

> canny_edged_image = load_image("https://huggingface.co/takuma104/controlnet_dev/resolve/main/vermeer_canny_edged.png")

Reviewer: ```suggestion
canny_edged_image = load_image("https://hf.co/datasets/huggingface/documentation-images/resolve/main/diffusers/vermeer_canny_edged.png")
```

pipe = StableDiffusionControlNetPipeline.from_pretrained("takuma104/control_sd15_canny").to("cuda")
image = pipe(prompt="best quality, extremely detailed", image=canny_edged_image).images[0]
image.save("generated.png")
```

Note that the text prompt does not make any reference to the structure or contents of the image we are generating. Stable Diffusion interprets the control image as an additional input that controls what to generate.

> - Controlling custom Stable Diffusion 1.5 models

Reviewer: Maybe you meant it as a heading?

> In the following example we use PromptHero's [Openjourney model](https://huggingface.co/prompthero/openjourney), which was fine-tuned from the base Stable Diffusion v1.5 model on images from Midjourney. This model has the same structure as Stable Diffusion 1.5 but is capable of producing a different output style.

Reviewer: ```suggestion
In the following example we use PromptHero's [Openjourney model](https://huggingface.co/prompthero/openjourney), which was fine-tuned from the base Stable Diffusion v1.5 model on images from Midjourney. This model has the same structure as Stable Diffusion 1.5 but is capable of producing outputs in a different style.
```

```py
from diffusers import StableDiffusionControlNetPipeline, AutoencoderKL, UNet2DConditionModel
from diffusers.utils import load_image

# Canny edged image for control

> canny_edged_image = load_image("https://huggingface.co/takuma104/controlnet_dev/resolve/main/vermeer_canny_edged.png")

Reviewer: ```suggestion
canny_edged_image = load_image("https://hf.co/datasets/huggingface/documentation-images/resolve/main/diffusers/vermeer_canny_edged.png")
```

Author: Fixed in https://github.com/huggingface/diffusers/pull/2407/commits/06bb1db313d1e3b87d82d662ae02245a6894f30f

base_model_id = "prompthero/openjourney"  # an example: openjourney model
vae = AutoencoderKL.from_pretrained(base_model_id, subfolder="vae").to("cuda")
unet = UNet2DConditionModel.from_pretrained(base_model_id, subfolder="unet").to("cuda")

pipe = StableDiffusionControlNetPipeline.from_pretrained("takuma104/control_sd15_canny", unet=unet, vae=vae).to("cuda")
image = pipe(prompt="best quality, extremely detailed", image=canny_edged_image, width=512, height=512).images[0]
image.save("generated.png")
```

[[autodoc]] StableDiffusionControlNetPipeline
	- all
	- __call__

> 	- enable_attention_slicing
> 	- disable_attention_slicing
> 	- enable_vae_slicing
> 	- disable_vae_slicing
> 	- enable_xformers_memory_efficient_attention
> 	- disable_xformers_memory_efficient_attention

Reviewer: We can discard these. `all` should take care of them :) 

```suggestion
```

Author: Fixed in https://github.com/huggingface/diffusers/pull/2407/commits/06bb1db313d1e3b87d82d662ae02245a6894f30f

Using the pretrained models we can provide control images (for example, a depth map) to control Stable Diffusion text-to-image generation so that it follows the structure of the depth image and fills in the details.

The abstract of the paper is the following:

*We present a neural network structure, ControlNet, to control pretrained large diffusion models to support additional input conditions. The ControlNet learns task-specific conditions in an end-to-end way, and the learning is robust even when the training dataset is small (< 50k). Moreover, training a ControlNet is as fast as fine-tuning a diffusion model, and the model can be trained on a personal devices. Alternatively, if powerful computation clusters are available, the model can scale to large amounts (millions to billions) of data. We report that large diffusion models like Stable Diffusion can be augmented with ControlNets to enable conditional inputs like edge maps, segmentation maps, keypoints, etc. This may enrich the methods to control large diffusion models and further facilitate related applications.*

> This model was contributed by the amazing community contributor [takuma104](https://huggingface.co/takuma104) ❤️ .

Reviewer 2: @takuma104 - left a comment here for future readers for credit

Author: @patrickvonplaten Thank you so much! It is a great honor!
